processor/worker.go

```diff
@@ -5,41 +5,163 @@ import (
 	"sync"
 	"time"
 
+	"golang.org/x/sync/errgroup"
+
 	"github.com/rudderlabs/rudder-go-kit/logger"
 	"github.com/rudderlabs/rudder-server/jobsdb"
 	"github.com/rudderlabs/rudder-server/rruntime"
 	"github.com/rudderlabs/rudder-server/services/rsources"
 	"github.com/rudderlabs/rudder-server/utils/misc"
 )
 
-// newProcessorWorker creates a new worker
+type worker struct {
+	partition string
+	workers   []*partitionWorker
+	logger    logger.Logger
+	stats     *processorStats
+	handle    workerHandle
+	g         *errgroup.Group
+}
+
+// newProcessorWorker creates a new worker for the specified partition
 func newProcessorWorker(partition string, h workerHandle) *worker {
 	w := &worker{
+		partition: partition,
+		logger:    h.logger().Child(partition),
+		stats:     h.stats(),
+		handle:    h,
+	}
+	w.g, _ = errgroup.WithContext(context.Background())
+
+	// Create workers for each partition
+	numPartitions := h.config().numPartitions
+	w.workers = make([]*partitionWorker, numPartitions)
+	for i := 0; i < numPartitions; i++ {
+		w.workers[i] = newWorker(partition, h)
+	}
+
+	return w
+}
+
+// Work processes jobs for the specified partition
+// Returns true if work was done, false otherwise
+func (w *worker) Work() bool {
+	// If pipelining is disabled, use the legacy job handling path
+	if !w.handle.config().enablePipelining {
+		return w.handle.handlePendingGatewayJobs(w.partition)
+	}
+
+	// Get jobs for this partition
+	jobs := w.handle.getJobs(w.partition)
+	start := time.Now()
+
+	// If no jobs were found, return false
+	if len(jobs.Jobs) == 0 {
+		return false
+	}
+
+	// Mark jobs as executing
+	if err := w.handle.markExecuting(w.partition, jobs.Jobs); err != nil {
+		w.logger.Error("Error marking jobs as executing", "error", err)
+		panic(err)
+	}
+
+	// Record throughput metrics
+	w.handle.stats().DBReadThroughput(w.partition).Count(throughputPerSecond(jobs.EventsCount, time.Since(start)))
+
+	// Initialize rsources stats
+	rsourcesStats := rsources.NewStatsCollector(w.handle.rsourcesService(), rsources.IgnoreDestinationID())
```

Comment:
Don't we need different stats per pipeline? Each pipeline should be independent, right?

Reply:
Not true, they are not being processed at the same time & as it stands now:
- statsCollector is not thread safe
- statsCollector's publish will be called multiple times from each pipeline with unpredictable results

Each set of jobs should have its own stats collector